domains/client/domain-operator/src/domain_block_processor.rs

```diff
@@ -547,89 +547,86 @@ where
-    //             Ordering::Equal => unreachable!("checked in this block's if condition; qed"),
-    //         }
-    //     }
-    // }
-    // match local_receipt
-    //     .invalid_bundles
-    //     .len()
-    //     .cmp(&external_receipt.invalid_bundles.len())
-    // {
-    //     Ordering::Greater => {
-    //         let invalid_bundle =
-    //             &local_receipt.invalid_bundles[external_receipt.invalid_bundles.len()];
-    //         Some(ReceiptMismatchInfo::InvalidBundles {
-    //             mismatch_type: InvalidBundlesMismatchType::InvalidAsValid,
-    //             bundle_index: invalid_bundle.bundle_index,
-    //             consensus_block_hash: local_receipt.consensus_block_hash
-    //         })
-    //     }
-    //     Ordering::Less => {
-    //         let valid_bundle =
-    //             &external_receipt.invalid_bundles[local_receipt.invalid_bundles.len()];
-    //         Some(ReceiptMismatchInfo::InvalidBundles {
-    //             mismatch_type: InvalidBundlesMismatchType::ValidAsInvalid,
-    //             bundle_index: valid_bundle.bundle_index,
-    //             consensus_block_hash: local_receipt.consensus_block_hash
-    //         })
-    //     }
-    //     Ordering::Equal => unreachable!("already checked for vector equality and since the zipped elements are equal, length cannot be equal; qed"),
-    // }
-    None
+    if local_receipt.bundles == external_receipt.bundles {
+        return Ok(None);
+    }
+
+    // The `bundles_extrinsics_roots` should be checked in the runtime when the consensus block is
+    // constructed/imported thus the `external_receipt` must have the same `bundles_extrinsics_roots`
+    //
+    // NOTE: this also check `local_receipt.bundles` and `external_receipt.bundles` have the same length
+    if local_receipt.bundles_extrinsics_roots() != external_receipt.bundles_extrinsics_roots() {
+        return Err(sp_blockchain::Error::Application(format!(
+            "Found mismatch of `ER::bundles_extrinsics_roots`, this should not happen, local: {:?}, external: {:?}",
+            local_receipt.bundles_extrinsics_roots(),
+            external_receipt.bundles_extrinsics_roots(),
+        ).into()));
+    }
+
+    // Get the first mismatch of `ER::bundles`
+    let (bundle_index, (local_bundle, external_bundle)) = local_receipt
+        .bundles
+        .iter()
+        .zip(external_receipt.bundles.iter())
+        .enumerate()
+        .find(|(_, (local_bundle, external_bundle))| local_bundle != external_bundle)
+        .expect(
+            "The `local_receipt.bundles` and `external_receipt.bundles` are checked to have the \
+            same length and being non-equal; qed",
+        );
+
+    // The `local_bundle` and `external_bundle` are checked to have the same `extrinsic_root`
+    // thus they must being mismatch due to bundle validity
+    let mismatch_type = match (local_bundle.bundle.clone(), external_bundle.bundle.clone()) {
+        (
```

Comment:
Please do not use nestings like. I have no clue how to process this function. Would be ideal if fucntionality to check a single bundle mismatch would be easier and does not hurt readability. Also Git diff from commented code to uncommented code is not helping either

Reply:
> Please do not use nestings like. I have no clue how to process this function. Would be ideal if fucntionality to check a single bundle mismatch would be easier and does not hurt readability.

Could you be more specific? I use `match` to handle each mismatch case in a dedicated branch which should be straightforward and clean IMHO, commented/uncommented code could hurt readability when reviewing though.